### PR TITLE
feat(pm_auth): Support different pm types in PM auth

### PR DESCRIPTION
**Pull request by Sarthak1799**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Support for SEPA and BACS in direct debit PM auth.
Earlier we only supported ACH based response from pm auth connector in PM auth, now have added support for BACS and SEPA based responses as well.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Have tested locally by hardcoding response from connector. This cannot be tested as it is in sandbox because Plaid always sends a basic ACH based response in sandbox which won't trigger the response structure for BACS and SEPA (the implementation of this PR).
A simple sanity testing however can be done for the feature, refer to this for testing instructions - https://github.com/juspay/hyperswitch/pull/3047

image
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/pm_auth/src/types.rs [hunk 1]
```diff
@@ -72,13 +72,36 @@ pub struct BankAccountCredentialsResponse {
...
     pub payment_method_type: PaymentMethodType,
     pub account_id: String,
     pub account_type: Option<String>,
 }
 
+#[derive(Debug, Clone)]
+pub enum PaymentMethodTypeDetails {
+    ACH(BankAccountDetailsACH),
+    BACS(BankAccountDetailsBACS),
+    SEPA(BankAccountDetailsSEPA),
```
Comment (vspecky) [inline crates/pm_auth/src/types.rs:85]:
```suggestion
    Ach(BankAccountDetailsAch),
    Bacs(BankAccountDetailsBacs),
    Sepa(BankAccountDetailsSepa),
```

**Review by vspecky — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payment_methods/cards.rs [hunk 2]
```diff
@@ -2831,12 +2831,20 @@ async fn get_bank_account_connector_details(
...
-                Ok(Some(BankAccountConnectorDetails {
-                    connector: connector_details.connector.clone(),
-                    account_id: connector_details.account_id.clone(),
-                    mca_id: connector_details.mca_id.clone(),
-                    access_token: connector_details.access_token.clone(),
-                }))
+
+                let pm_type = pm
+                    .payment_method_type
+                    .ok_or(errors::ApiErrorResponse::InternalServerError)
```
Comment (vspecky) [inline crates/router/src/core/payment_methods/cards.rs:2837]:
```suggestion
                    .get_required_value("payment_method_type")
```

File: crates/router/src/core/pm_auth.rs [hunk 3]
```diff
@@ -719,20 +741,41 @@ pub async fn retrieve_payment_method_from_auth_service(
...
-        account_number: masking::Secret::new(bank_account.account_number.clone()),
-        routing_number: masking::Secret::new(bank_account.routing_number.clone()),
-        card_holder_name: None,
-        bank_account_holder_name: None,
-        bank_name: None,
-        bank_type,
-        bank_holder_type: None,
-    });
+    let billing_details = BankDebitBilling {
+        name: name.unwrap_or_default(),
```
Comment (vspecky) [inline crates/router/src/core/pm_auth.rs:745]: Don't use `unwrap_or_default()`. Keep it a sensible default or keep it `None` if absent.

**Review by vspecky — CHANGES_REQUESTED** [review]

File: crates/router/src/core/pm_auth.rs [hunk 4]
```diff
@@ -357,7 +357,24 @@ async fn store_bank_details_in_payment_methods(
     let mut new_entries: Vec<storage::PaymentMethodNew> = Vec::new();
 
     for creds in bank_account_details_resp.credentials {
-        let hash_string = format!("{}-{}", creds.account_number, creds.routing_number);
+        let (account_number, hash_string) = match creds.account_details {
+            pm_auth_types::PaymentMethodTypeDetails::Ach(ach) => (
+                ach.account_number.clone(),
+                format!(
+                    "{}-{}",
```
Comment (vspecky) [inline crates/router/src/core/pm_auth.rs:364]: Lets include additional metadata like the payment method type (Bacs, Sepa), in the hash.

**Review by vspecky — CHANGES_REQUESTED** [review]

**Review by vspecky — APPROVED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Pull request merged**